util/dao/src/lib.rs

```diff
@@ -1,48 +1,267 @@
+                                self.calculate_maximum_withdraw(
+                                    &output,
+                                    &deposit_header_hash,
+                                    &withdraw_header_hash,
+                                )
+                            } else {
+                                Ok(output.capacity)
+                            }
+                        }
+                    };
+                    capacity.and_then(|c| c.safe_add(capacities).map_err(Into::into))
+                },
+            )
+            .and_then(|x| {
+                rtx.transaction
+                    .outputs_capacity()
+                    .and_then(|y| x.safe_sub(y))
+                    .map_err(Into::into)
+            })
+    }
+
+    fn input_occupied_capacities(
+        &self,
+        rtx: &ResolvedTransaction,
+    ) -> Result<Capacity, FailureError> {
+        rtx.resolved_inputs
+            .iter()
+            .try_fold(Capacity::zero(), |capacities, resolved_input| {
+                let current_capacity = if let Some(cell_meta) = resolved_input.cell() {
+                    let output = self.data_loader.lazy_load_cell_output(&cell_meta);
+                    output.occupied_capacity()
+                } else {
+                    Ok(Capacity::zero())
+                };
+                current_capacity.and_then(|c| capacities.safe_add(c))
+            })
+            .map_err(Into::into)
+    }
 
-    let withdraw_counted_capacity = u128::from(counted_capacity.as_u64())
-        * u128::from(withdraw_dao_stats.accumulated_rate)
-        / u128::from(deposit_dao_stats.accumulated_rate);
+    fn calculate_maximum_withdraw(
+        &self,
+        output: &CellOutput,
+        deposit_header_hash: &H256,
+        withdraw_header_hash: &H256,
+    ) -> Result<Capacity, FailureError> {
+        let deposit_header = self
+            .store
+            .get_block_header(deposit_header_hash)
+            .ok_or(Error::InvalidHeader)?;
+        let withdraw_header = self
+            .store
+            .get_block_header(withdraw_header_hash)
+            .ok_or(Error::InvalidHeader)?;
+        let (deposit_ar, _, _) = extract_dao_data(deposit_header.dao())?;
+        let (withdraw_ar, _, _) = extract_dao_data(withdraw_header.dao())?;
 
-    let withdraw_capacity =
-        Capacity::shannons(withdraw_counted_capacity as u64).safe_add(occupied_capacity)?;
-    Ok(withdraw_capacity)
+        let occupied_capacity = output.occupied_capacity()?;
+        let counted_capacity = output.capacity.safe_sub(occupied_capacity)?;
+        let withdraw_counted_capacity = u128::from(counted_capacity.as_u64())
+            * u128::from(withdraw_ar)
+            / u128::from(deposit_ar);
+        let withdraw_capacity =
+            Capacity::shannons(withdraw_counted_capacity as u64).safe_add(occupied_capacity)?;
+
+        Ok(withdraw_capacity)
+    }
 }
 
 fn calculate_g2(
```

Comment:
Should we make it consistent with block reward? A.k.a, the reminder is split evenly to the first r blocks, where r is the remainder.

Reply:
Since we cannot issue reward to genesis, we should exclude genesis from g2 accumulation. A.k.a, g2 of genesis block is 0.

- G2: Secondary Reward in an Epoch
- `r_i`: Secondary rate of block i

| Block | AR |
| ----- | ---- |
| Genesis | `1` |
| Block 1 | `1`  |
| Block 2 | `1 * (1 + r_1)` |
| Block 3 | `1 * (1 + r_1) + (1 + r_2)` |